lib/conditioned.rb

```diff
@@ -221,9 +222,10 @@ def maxHeatScheduledSetpoint(zone)
   # standards/Standards.ThermalZone.rb#L910
   setpoint = nil
   dual = false
-  raise "Invalid zone (max T)" unless zone
-  cl = OpenStudio::Model::ThermalZone
-  raise "#{zone.class}? expected #{cl} (max T)" unless zone.is_a?(cl)
+  unless zone && zone.is_a?(OpenStudio::Model::ThermalZone)
+    TBD.log(TBD::DEBUG, "Invalid zone - max T")
```

Comment:
The 'max T' here would confuse me if I was a user, I would wonder I do something wrong with a max T?  Really something else is going on here, it seems like you really should never get to this line in your code.  Is it really a debug level?

For your purposes you want a unique string so you can see which line of code failed.  You could add some code in your logger to get the line of code where the log statement was called using `caller`

https://stackoverflow.com/questions/6322524/how-do-you-view-a-sample-of-the-call-stack-in-ruby

Reply:
I mostly agree here. First things first: as mentioned earlier, I'm currently revising the TBD log messages to be more informative (within reason) - in this case, likely reverting back to the original checks (e.g., invalid/empty _zone_ object, wrong class). Certainly more practical for 3rd-party developers relying on TBD. This addresses part of your suggestion. Otherwise, I wasn't planning to set up a _production debugging_ solution for TBD (this would require an additional _TBD Measure_ 'debug' argument, I presume). TBD DEBUG messages should not make their way in production mode. _Normal_ users (e.g., designers, modellers) would not have the option of running TBD in DEBUG mode (unless reassembling TBD), and therefore wouldn't come across these (and so shouldn't be confused). As with most errors labelled under DEBUG (in lower-level functions), TBD simply keeps track of safe _exits_ preventing Ruby crashes. For such an error to happen, something has to have gone horribly wrong upstream e.g., an invalid OSM that wasn't caught by OpenStudio (and/or the Measures code) or an internal TBD bug. If one navigates the TBD calling code for _maxHeatScheduledSetpoint_, adequate safeguards are in place to catch these. I would rather fix the upstream code then rely on lower-level functions to report invalid `.osm` file entries. If we ever felt it necessary to truly validate OSM entries beforehand, I would prefer setting up a _validOSM?_ function that would log everything wrong with an OSM from TBD/Topolys' POV - a lot of effort. Should we be parsing OpenStudio logs for such anomalies? For instance, a 5-sided window should be red-flagged by OpenStudio, IMO - is this the case?

Reply:
I think a verify OSM method would be good before doing any TDB processing, of course that can wait until you have time/resources to work on it.  My one input would be that somethings you think of as OSM errors (e.g. 5-sided window) are not necessarily going to be flagged by OpenStudio.  It probably is going to be up to TBD to verify that the OSM conforms to any TBD specific requirements.

Reply:
Got it. Just to be clear for others: TBD/Topolys does not have particular requirements concerning many of these errors - EnergyPlus does. The [5-sided window](https://github.com/NREL/EnergyPlus/issues/6534) case is a good example: TBD/Topolys can handle n-sided, irregular, concave vs convex subsurfaces just fine. As it stands, OpenStudio relies on EnergyPlus to raise the 5-sided window error and halt the simulation. So be it, but it does make writing Measures more challenging as many of these errors need to raised pre-simulation. For OSM geometry/topolgy, this would not be limited to TBD/Topolys (e.g. cost estimation, AFN). So taking the `validOSM?` function (or group of functions) a step further - maybe better to extend the [OpenStudio Common Measures Gem](https://github.com/macumber/openstudio-common-measures-gem/tree/develop/lib/measures)? I'd prefer putting (future) resources on that, and have TBD either recommend users rely on a handful of those - or have TBD integrate a number of these directly (under `/resources/`). That would lighten the load for TBD and make it (more readily) available for other Measures or Ruby scripts.